src/metatensor/models/utils/data/combine_dataloaders.py

```diff
@@ -1,57 +1,38 @@
-import itertools
 from typing import List
 
 import numpy as np
 import torch
 
 
-class CombinedIterableDataset(torch.utils.data.IterableDataset):
-    """
-    Combines multiple dataloaders into a single iterable dataset.
-    This is useful for combining multiple dataloaders into a single
-    dataloader. The new dataloader can be shuffled or not.
-
-    :param dataloaders: list of dataloaders to combine
-    :param shuffle: whether to shuffle the combined dataloader
-
-    :return: combined dataloader
-    """
-
-    def __init__(self, dataloaders, shuffle):
+class CombinedDataLoader:
+    def __init__(self, dataloaders: List[torch.utils.data.DataLoader], shuffle: bool):
         self.dataloaders = dataloaders
         self.shuffle = shuffle
 
         # Create the indices:
-        indices = [
-            (i, dl_idx)
-            for dl_idx, dl in enumerate(self.dataloaders)
-            for i in range(len(dl))
-        ]
+        self.indices = list(range(len(self)))
 
         # Shuffle the indices if requested
         if self.shuffle:
-            np.random.shuffle(indices)
-
-        self.indices = indices
+            np.random.shuffle(self.indices)
 
-    def __iter__(self):
-        for idx, dataloader_idx in self.indices:
-            yield next(itertools.islice(self.dataloaders[dataloader_idx], idx, None))
+        self.reset()
 
-    def __len__(self):
-        return len(self.indices)
+    def reset(self):
+        self.current_index = 0
+        self.full_list = [batch for dl in self.dataloaders for batch in dl]
 
+    def __iter__(self):
+        return self
```

Comment:
Is this working? Shouldn't you return the next dataloader here instead of the full instance?

Reply:
It does work. It comes from ChatGPT, but intuitively it makes sense:
You call `iterable = iter(dataloader)` and then `next(iterable)` (a bunch of times). Next is defined in the class, so it makes sense to return self

Reply:
It's there already from the old function + class